### PR TITLE
INT-4204: Introduce `ExpressionFileListFilter`

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4204

### CODE REVIEW

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/ExpressionFileListFilter.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.util.Assert;
+
+/**
+ * A SpEL expression based {@link AbstractFileListFilter} implementation.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.0
+ */
+public class ExpressionFileListFilter<F> extends AbstractFileListFilter<F>
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/ExpressionFileListFilter.java]: Given that SpEL is not type safe, and we don't need subclasses, should this be generic?

    public class ExpressionFileListFilter extends AbstractFileListFilter<Object>

?

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/ExpressionFileListFilter.java (hunk 1 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/ExpressionFileListFilter.java]: Well, we could, indeed but then we can't use `AbstractInboundFileSynchronizer.setFilter(FileListFilter<F> filter)` as is.
We have to go through all the casting hell.

Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/ExpressionFileListFilter.java]: 👍

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Merged as 9836828f54def6e3276b3370d8e2dab9f9788020

**Pull request closed without merging**